File: v2.x/Examples/SciChartDemo/SciChartDemo/Views/CreateRealtimeCharts/RealtimeGhostTraces/RealTimeGhostTracesChartView.m
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2018. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// RealTimeGhostTracesChartView.m is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

#import "RealTimeGhostTracesChartView.h"
#import "DataManager.h"
#import "RandomUtil.h"

@implementation RealTimeGhostTracesChartView {
    NSTimer * _timer;

    double _lastAmplitude;
    double _phase;
}

- (void)commonInit {
    __weak typeof(self) wSelf = self;
    self.speedChanged = ^(UISlider * sender) { [wSelf onSpeedChanged:sender]; };
}

- (void)initExample {
    _lastAmplitude = 1.0;
    
    id<SCIAxis2DProtocol> xAxis = [SCINumericAxis new];
    xAxis.autoRange = SCIAutoRange_Always;
    
    id<SCIAxis2DProtocol> yAxis = [SCINumericAxis new];
    yAxis.autoRange = SCIAutoRange_Never;
    yAxis.visibleRange = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(-2.0) Max:SCIGeneric(2.0)];
    
    [self.surface.xAxes add:xAxis];
    [self.surface.yAxes add:yAxis];
    
    uint limeGreen = 0xFF32CD32;
    
    [self addLineRenderableSeries:limeGreen opacity:1.0];
    [self addLineRenderableSeries:limeGreen opacity:0.9];
    [self addLineRenderableSeries:limeGreen opacity:0.8];
    [self addLineRenderableSeries:limeGreen opacity:0.7];
    [self addLineRenderableSeries:limeGreen opacity:0.62];
    [self addLineRenderableSeries:limeGreen opacity:0.55];
    [self addLineRenderableSeries:limeGreen opacity:0.45];
    [self addLineRenderableSeries:limeGreen opacity:0.35];
    [self addLineRenderableSeries:limeGreen opacity:0.25];
    [self addLineRenderableSeries:limeGreen opacity:0.15];
    
    _timer = [NSTimer scheduledTimerWithTimeInterval:0.02 target:self selector:@selector(updateData:) userInfo:nil repeats:YES];
}

- (void)willMoveToWindow:(UIWindow *)newWindow{
    [super willMoveToWindow: newWindow];
    
    if (newWindow == nil) {
        [_timer invalidate];
        _timer = nil;
    }
}

- (void)onSpeedChanged:(UISlider *) sender {
    if (sender.value > 0) {
        if (_timer != nil) {
            [_timer invalidate];
        }
        _timer = [NSTimer scheduledTimerWithTimeInterval:sender.value / 1000.0 target:self selector:@selector(updateData:) userInfo:nil repeats:YES];
    }
}

- (void)updateData:(NSTimer *)timer {
    [SCIUpdateSuspender usingWithSuspendable:self.surface withBlock:^{
        SCIXyDataSeries * dataSeries = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Double YType:SCIDataType_Double];

        double randomAmplitude = _lastAmplitude + ([RandomUtil nextDouble] - 0.5);
        if (randomAmplitude < -2.0) {
            randomAmplitude = -2.0;
        } else if (randomAmplitude > 2.0) {
            randomAmplitude = 2.0;
        }
        
        DoubleSeries * noisySinewave = [DataManager getNoisySinewaveWithAmplitude:randomAmplitude Phase:_phase PointCount:1000 NoiseAmplitude:0.25];
        _lastAmplitude = randomAmplitude;
    
        [dataSeries appendRangeX:noisySinewave.xValues Y:noisySinewave.yValues Count:noisySinewave.size];
        
        [self reassignRenderSeriesWithDataSeries:dataSeries];
    }];
}

- (void)reassignRenderSeriesWithDataSeries:(SCIXyDataSeries *)dataSeries {
    SCIRenderableSeriesCollection * rs = self.surface.renderableSeries;
    
    // shift old dataSeries
    [rs itemAt:9].dataSeries = [rs itemAt:8].dataSeries;
    [rs itemAt:8].dataSeries = [rs itemAt:7].dataSeries;
    [rs itemAt:7].dataSeries = [rs itemAt:6].dataSeries;
    [rs itemAt:6].dataSeries = [rs itemAt:5].dataSeries;
    [rs itemAt:5].dataSeries = [rs itemAt:4].dataSeries;
    [rs itemAt:4].dataSeries = [rs itemAt:3].dataSeries;
    [rs itemAt:3].dataSeries = [rs itemAt:2].dataSeries;
    [rs itemAt:2].dataSeries = [rs itemAt:1].dataSeries;
    [rs itemAt:1].dataSeries = [rs itemAt:0].dataSeries;
    
    // use new dataSeries to draw first renderableSeries
    [rs itemAt:0].dataSeries = dataSeries;
}

- (void)addLineRenderableSeries:(uint)color opacity:(float)opacity {
    SCIFastLineRenderableSeries * lineRenerableSeries = [SCIFastLineRenderableSeries new];
    lineRenerableSeries.opacity = opacity;
    lineRenerableSeries.strokeStyle = [[SCISolidPenStyle alloc] initWithColorCode:color withThickness:1];
    
    [self.surface.renderableSeries add:lineRenerableSeries];
}

@end

File: v2.x/Examples/SciChartSwiftDemo/SciChartSwiftDemo/Views/CreateRealtimeCharts/RealtimeGhostTraces/RealTimeGhostTracesChartView.swift
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2018. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// RealTimeGhostTracesChartView.swift is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

class RealTimeGhostTracesChartView: RealtimeGhostTracesLayout {
    
    private var _timer: Timer?
    private var _lastAmplitude = 1.0
    private var _phase = 0.0
    
    override func commonInit() {
        weak var wSelf = self
        self.speedChanged = { sender in wSelf?.onSpeedChanged(sender!) }
    }
    
    override func initExample() {
        let xAxis = SCINumericAxis()
        xAxis.autoRange = .always
        
        let yAxis = SCINumericAxis()
        yAxis.autoRange = .never
        yAxis.visibleRange = SCIDoubleRange.init(min: SCIGeneric(-2), max: SCIGeneric(2))
        
        surface.xAxes.add(xAxis)
        surface.yAxes.add(yAxis)
        
        let limeGreen: uint = 0xFF32CD32
        
        addLineRenderableSeries(limeGreen, opacity: 1.0)
        addLineRenderableSeries(limeGreen, opacity: 0.9)
        addLineRenderableSeries(limeGreen, opacity: 0.8)
        addLineRenderableSeries(limeGreen, opacity: 0.7)
        addLineRenderableSeries(limeGreen, opacity: 0.62)
        addLineRenderableSeries(limeGreen, opacity: 0.55)
        addLineRenderableSeries(limeGreen, opacity: 0.45)
        addLineRenderableSeries(limeGreen, opacity: 0.35)
        addLineRenderableSeries(limeGreen, opacity: 0.25)
        addLineRenderableSeries(limeGreen, opacity: 0.15)

        _timer = Timer.scheduledTimer(timeInterval: 0.02, target: self, selector: #selector(updateData), userInfo: nil, repeats: true)
    }
    
    override func willMove(toWindow newWindow: UIWindow?) {
        super.willMove(toWindow: newWindow)
        
        if newWindow == nil {
            _timer?.invalidate()
            _timer = nil
        }
    }
    
    private func onSpeedChanged(_ sender:UISlider) {
        _timer?.invalidate()
        _timer = Timer.scheduledTimer(timeInterval: Double(sender.value) / 1000.0, target: self, selector: #selector(updateData), userInfo: nil, repeats: true)
    }
    
    @objc fileprivate func updateData(_ timer:Timer) {
        SCIUpdateSuspender.usingWithSuspendable(surface) {
            let dataSeries = SCIXyDataSeries.init(xType: .double, yType: .double)

            var randomAmplitude:Double = self._lastAmplitude + (RandomUtil.nextDouble() - 0.5)
            if (randomAmplitude < -2.0) {
                randomAmplitude = -2.0
            } else if (randomAmplitude > 2.0) {
                randomAmplitude = 2.0;
            }
            
            let noisySinewave = DataManager.getNoisySinewave(withAmplitude: randomAmplitude, phase: self._phase, pointCount: 1000, noiseAmplitude: 0.25)
            self._lastAmplitude = randomAmplitude;
            
            dataSeries.appendRangeX(noisySinewave!.xValues, y: noisySinewave!.yValues, count: noisySinewave!.size)
            self.reassignRenderSeriesWith(dataSeries: dataSeries)
        }
    }
    
    fileprivate func reassignRenderSeriesWith(dataSeries: SCIXyDataSeries) {
        let rs = surface.renderableSeries
        
        // shift old dataSeries
        rs.item(at: 9).dataSeries = rs.item(at: 8).dataSeries
        rs.item(at: 8).dataSeries = rs.item(at: 7).dataSeries
        rs.item(at: 7).dataSeries = rs.item(at: 6).dataSeries
        rs.item(at: 6).dataSeries = rs.item(at: 5).dataSeries
        rs.item(at: 5).dataSeries = rs.item(at: 4).dataSeries
        rs.item(at: 4).dataSeries = rs.item(at: 3).dataSeries
        rs.item(at: 3).dataSeries = rs.item(at: 2).dataSeries
        rs.item(at: 2).dataSeries = rs.item(at: 1).dataSeries
        rs.item(at: 1).dataSeries = rs.item(at: 0).dataSeries
        
        // use new dataSeries to draw first renderableSeries
        rs.item(at: 0).dataSeries = dataSeries
    }
    
    fileprivate func addLineRenderableSeries(_ colorCode: UInt32, opacity: Float){
        let lineRenderSeries = SCIFastLineRenderableSeries()
        lineRenderSeries.opacity = opacity
        lineRenderSeries.strokeStyle = SCISolidPenStyle.init(colorCode: colorCode, withThickness: 1)
        
        surface.renderableSeries.add(lineRenderSeries)
    }
}
